datarobot_batch_scoring/network.py

```diff
@@ -218,66 +286,110 @@ def __iter__(self):
+                                      ''.format(self.network_status.value))
+                        self.network_status.value = b"R"
+                        f = self._executor.submit(self._request, r)
+                        f.add_done_callback(self.request_cb)
+                        self.futures.append(f)
+                        break
+                    else:
+                        self.ui.debug('state: {} -> F'
+                                      ''.format(self.network_status.value))
+                        self.network_status.value = b"F"
+                        wait(self.futures, return_when=FIRST_COMPLETED)
+                yield
         #  wait for all batches to finish before returning
+        self.network_status.value = b"W"
         while self.futures:
             f_len = len(self.futures)
             self.futures = [i for i in self.futures if not i.done()]
             if f_len != len(self.futures):
-                self._ui.debug('Waiting for final requests to finish. '
-                               'remaining requests: {}'
-                               ''.format(len(self.futures)))
+                self.ui.debug('Waiting for final requests to finish. '
+                              'remaining requests: {}'
+                              ''.format(len(self.futures)))
             wait(self.futures, return_when=FIRST_COMPLETED)
+        self.network_status.value = b"D"
         yield True
 
     def __enter__(self):
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
-        self._executor.shutdown(wait=True)
+        if self.proc and self.proc.is_alive():
+            self.proc.terminate()
+
+    def run(self, dry_run=False):
+        if dry_run:
+            i = 0
+            for _ in self.perform_requests(True):
+                i += 1
+
+            return i
+
+        self._executor = ThreadPoolExecutor(self.concurrency)
+        self.session = requests.Session()
+        adapter = requests.adapters.HTTPAdapter(
+            pool_connections=self.concurrency, pool_maxsize=self.concurrency)
+        self.session.mount('http://', adapter)
+        self.session.mount('https://', adapter)
+
+        t0 = time()
+        i = 0
+        r = None
+        for r in self.perform_requests():
+            i += 1
+            self.ui.info('{} responses sent | time elapsed {}s'
+                         .format(i, time() - t0))
+
+        if r is True:
+            i -= 1
+
+        self.progress_queue.put((ProgressQueueMsg.NETWORK_DONE, {
+            "ret": r,
+            "processed": self.n_requests,
+            "retried": self.n_retried,
+            "consumed": self.n_consumed,
+            "rusage": get_rusage(),
+        }))
+
+    def go(self, dry_run=False):
+        if dry_run:
+            return self.run(dry_run)
+
+        self.ui.set_next_UI_name('network')
+        self.proc = \
+            multiprocessing.Process(target=self.run,
```

Comment:
I think you misunderstood what I was saying. I was trying to say is that I do not think that we need a new process just for compression...yet. I don't think the overhead for compression is that large.  I will leave it up to you.

The resource info (which is very cool!) does show we spend a not-insignificant amount of time compressing. This is on a 1G dataset with a fast connection, 31 workers and a fast model. The bottleneck is the writer in this case. 
```
# NO COMPRESS 
MainProcess [INFO] Shovel is finished ok. Chunks produced: 409 Resource usage:                   { 'stime': 2.499195, 'utime': 12.081455, 'rss': 261308}
MainProcess [INFO] Network is finished ok. Chunks: 409 Requests: 409 Retries: 0 Resource usage:  { 'stime': 5.085765, 'utime': 23.361974, 'rss': 704068,}
MainProcess [INFO] Writer is finished ok. Result: True Results: 409 Written: 409 Resource usage: { 'stime': 1.429178, 'utime': 52.195045, 'rss': 69672,}
MainProcess [INFO] scoring complete | time elapsed 61.48334240913391s
MainProcess [INFO] scoring complete | total time elapsed 61.92390727996826s

# WITH COMPRESS 
MainProcess [INFO] Shovel is finished ok. Chunks produced: 409 Resource usage:                   {'stime': 2.523238, 'utime': 11.233651, 'rss': 260676}
MainProcess [INFO] Network is finished ok. Chunks: 409 Requests: 409 Retries: 0 Resource usage:  {'stime': 4.524693, 'utime': 45.154097, 'rss': 650000}
MainProcess [INFO] Writer is finished ok. Result: True Results: 409 Written: 409 Resource usage: {'stime': 1.59124,  'utime': 54.765873, 'rss': 68696}
MainProcess [INFO] scoring complete | time elapsed 63.556153535842896s
MainProcess [INFO] scoring complete | total time elapsed 63.95048785209656s
```
23 vs 45 seconds.